compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -1654,7 +1654,20 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                 match *cast_kind {
                     CastKind::PointerCoercion(PointerCoercion::ReifyFnPointer, coercion_source) => {
                         let is_implicit_coercion = coercion_source == CoercionSource::Implicit;
-                        let src_sig = op.ty(body, tcx).fn_sig(tcx);
+                        let src_ty = op.ty(body, tcx);
+                        let mut src_sig = src_ty.fn_sig(tcx);
+                        if let ty::FnDef(def_id, _) = src_ty.kind()
+                            && let ty::FnPtr(_, target_hdr) = *ty.kind()
+                            && tcx.codegen_fn_attrs(def_id).safe_target_features
+                            && target_hdr.safety.is_safe()
```

Comment:
these two lines can go into `adjust_target_feature_sig`, too

Reply:
wait, the first one already is lol

Reply:
(it's not?)
I thought about it, and I am not convinced it would be a good idea - in coercion, this check needs to be separate, as otherwise we'd emit the wrong error.
We could of course check twice in the coercion codepath, but then it's not really deduplicating...

Reply:
ah right. here we can in fact just unwrap the option as it should be unreachable due to the error emitted in typeck tainting the entire body